src/ev3_bluetooth.c

```diff
@@ -0,0 +1,332 @@
+         * interrupt, we should be able to call 'connect'
+         * again, but for some reason if we do it errno becomes EBADFD.
+         * 
+         * We didn't ask to use the socket in a non blocking wait (using the SOCK_NONBLOCK flag), but if we check the connection
+         * status with 'pool' and 'getsockopt' (the way we would do if we were using async socket) it works.
+        */
+        if (connectResult == 0 || errno == EINTR) {
+            bool connected = waitAsyncConnection(socketToServer);
+            if (connected) {
+                return socketToServer;
+            }
+        }
+        
+        // try to connect again later
+        close(socketToServer);
+        sleep(1);
+        attempts--;
+    }
+    return -1;
+}
+
+bool waitAsyncConnection (int socket) {
+    struct pollfd toMonitor;
+    toMonitor.fd = socket;
+    toMonitor.events = POLLOUT;
+    int poolResult;
+    do {
+        poolResult = poll(&toMonitor, 1, -1);
+    } while (poolResult == -1 && errno == EINTR);
+
+    if (poolResult != -1) {
+        int optval;
+        socklen_t optLength = sizeof(optval);
+        int getSockOptResult = getsockopt(socket, SOL_SOCKET, SO_ERROR, &optval, &optLength);
+        if (getSockOptResult == 0 && optval == 0) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+bool isServerStarted();
+int startServer ();
+BluetoothConnectionHandle acceptIncomingConnection ();
+
+BluetoothConnectionHandle WaitConnection() {
+	if(!isServerStarted()) {
+	    if(startServer() != 0) {
+	        return -1;
+	    }
+	}
+	return acceptIncomingConnection();
+}
+
+
+int serverSocket = -1;
+
+bool isServerStarted(){
+	return serverSocket > 0;
+}
+
+int bindServerSocket ();
+
+int startServer () {
+    serverSocket = socket(AF_BLUETOOTH, SOCK_STREAM, BTPROTO_RFCOMM);
+    if (serverSocket == -1) {
+        return -1;
+    }
+	if(bindServerSocket() != 0) {
+	    return -1;
+	}
+    return listen(serverSocket, 1); // TODO: do we keep 1 as backlog?
+}
+
+int bindServerSocket () {
+    struct sockaddr_rc listenAddress;
+    listenAddress.rc_family = AF_BLUETOOTH;
+    listenAddress.rc_bdaddr = *BDADDR_ANY;
+    listenAddress.rc_channel = DEFAULT_C4EV3_RFCOMM_CHANNEL;
+    return bind(serverSocket, (struct sockaddr *) &listenAddress, sizeof(listenAddress));
+}
+
+BluetoothConnectionHandle acceptIncomingConnection () {
+	struct sockaddr_rc remoteAddress = { 0 };
+	socklen_t addressLength = sizeof(remoteAddress);
+    BluetoothConnectionHandle c;
+    do {
+    	c = accept(serverSocket, (struct sockaddr *) &remoteAddress, &addressLength);
+    } while (c == -1 && errno == EINTR);
+    return c;
+}
+
+
+int SendStringTo(BluetoothConnectionHandle to, const char * str) {
+	return write(to, str, strlen(str));
+}
+
+int ReceiveStringFrom(BluetoothConnectionHandle from, char * buffer, int bufferLength) {
+	int readBytes;
+	do {
+		readBytes = read(from, buffer, bufferLength - 1);
```

Comment:
Can you have short reads?

Reply:
Yes and this is intended, see https://github.com/simonedegiacomi/EV3-API/commit/64e12eaa8a0903cd97b656a015485862630e1e35#r37078805